src/algorithms/dqns/qr_dqn.jl

```diff
@@ -0,0 +1,169 @@
+    batch_size::Int = 32,
+    update_horizon::Int = 1,
+    min_replay_history::Int = 32,
+    update_freq::Int = 1,
+    quantile_num::Int = 1,
+    target_update_freq::Int = 100,
+    traces = SARTS,
+    update_step = 0,
+    κ::Float32 = 1.0f0,
+    rng = Random.GLOBAL_RNG,
+) where {Tq,Tt,Tf}
+    copyto!(approximator, target_approximator)
+    sampler = NStepBatchSampler{traces}(;
+        γ = γ,
+        n = update_horizon,
+        stack_size = stack_size,
+        batch_size = batch_size,
+    )
+
+    N = quantile_num
+    τ = collect(0:N+1)/N
+    τ  = (τ[2:N+1]+τ[1:N])/2
+    τ = reshape(τ, 1, N)
+
+    QRDQNLearner(
+        approximator,
+        target_approximator,
+        loss_func,
+        min_replay_history,
+        update_freq,
+        update_step,
+        target_update_freq,
+        sampler,
+        N,
+        rng,
+        κ,
+        τ,
+        0.0f0,
+    )
+end
+
+Flux.functor(x::QRDQNLearner) = (Q = x.approximator, Qₜ = x.target_approximator),
+y -> begin
+    x = @set x.approximator = y.Q
+    x = @set x.target_approximator = y.Qₜ
+    x
+end
+
+function (learner::QRDQNLearner)(env)
+    s = send_to_device(device(learner.approximator), state(env))
+    s = Flux.unsqueeze(s, ndims(s) + 1)
+    q = reshape(learner.approximator(s), learner.quantile_num, :)
+    vec(mean(q, dims = 1)) |> send_to_host
+end
+
+function quantile_huber_loss(target::AbstractArray, q::AbstractArray, κ::Float32, τ::AbstractArray, N::Int)
+    TD_error = (target .- q)
+
+    # dropgrad
+    temp = Zygote.dropgrad(abs.(TD_error) .<  κ)
+    element_wise_huber_loss = ((TD_error.^2) .* temp) + κ*(TD_error .- 0.5*κ) .* (1 .- temp)
+    element_wise_huber_loss =
+        abs.(τ .- Zygote.dropgrad(TD_error .< 0)) .*
+        element_wise_huber_loss ./ κ
+
+    batch_quantile_huber_loss = mean(sum(element_wise_huber_loss; dims = 1), dims=1)
+    quantile_huber_loss =
+        mean(batch_quantile_huber_loss)
+
+    quantile_huber_loss
+    end
+
+function RLBase.update!(learner::QRDQNLearner, batch::NamedTuple)
+    Q = learner.approximator
+    Qₜ = learner.target_approximator
+    γ = learner.sampler.γ
+    loss_func = learner.loss_func
+    n = learner.sampler.n
+    batch_size = learner.sampler.batch_size
+    N = learner.quantile_num
+    κ = learner.κ
+    τ = learner.τ
+    D = device(Q)
+
+    s, a, r, t, s′ = (send_to_device(D, batch[x]) for x in SARTS)
+    a = CartesianIndex.(repeat(batch.action, inner = N), 1:(N*batch_size))
+
+    target_q = reshape(Qₜ(s′), N, :, batch_size)
+    avg_q = mean(target_q, dims=1)
+
+    if haskey(batch, :next_legal_actions_mask)
+        l′ = send_to_device(D, batch[:next_legal_actions_mask])
+        avg_q .+= ifelse.(l′, 0.0f0, typemin(Float32))
+    end
+
+    aₜ = argmax(avg_q, dims=2)
+    aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:N-1, 0:0, 0:0)))
+    qₜ = reshape(view(target_q, aₜ), :, batch_size)
+    target =
+    reshape(r, 1, batch_size) .+ γ * reshape(1 .- t, 1, batch_size) .* qₜ
```

Comment:
The calculation of target quantiles seems to be incorrect here. Please have a check.

Reply:
I have checked the code and the outputs with jupyter notebook and the selection and calculation of values seem to be right to me. 

If there is anything specific, please do let me know.

Reply:
OK, I was meant to let you figure it out by yourself. Could you share more details about how you check the code and outputs? 

From the algorithm you pasted above:

![image](https://user-images.githubusercontent.com/5612003/116034117-5321e400-a695-11eb-8e96-fafb2bebd1f6.png)

But obviously you are not calculating the correct quantile value here. This makes me suspect that if you have fully understood the algorithm before implementing it.

Reply:
I am sorry. I wasn't able to understand much of the notation given in the paper so I used https://github.com/ku2482/fqf-iqn-qrdqn.pytorch for reference. 

I just made sure that the code worked as intended by getting the outputs and the outputs are as expected.

Clearly, there is something wrong with the approach. I will try to fix that.

Reply:
Here, the `j` means the `j`-th quantile, and `a*` is the selected best action.

So to calculate the `target_quantile`, we simply select the the `a*`-th slice along the second dimension of `qₜ`.

Reply:
In the implemented algorithm:

- I take the average for all the quantiles for the actions (1st dimension) for s'.
- Then take the argmax to find the best action. 
- Then found the target slices indices using at.
- Then used at to select the action slices in Q(s').
- Then calculated target.

This sounds the same as the interpretation of the algorithm.

I am sorry if I got something wrong.

Reply:
`aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:N-1, 0:0, 0:0)))` I guess this line doesn't work as you expected.

Reply:
**Works as expected:**
![image](https://user-images.githubusercontent.com/49596933/116082057-980b4200-a6b8-11eb-9e0f-3dcfadf84cdd.png)
![image](https://user-images.githubusercontent.com/49596933/116082159-b8d39780-a6b8-11eb-9387-a1f7cc581bbb.png)

Reply:
Oh, alright. That makes sense. 

I think what you suggest is more efficient. So, can I make a change?